src/cookbooks/bigquery/optimization.md

```diff
@@ -1,28 +1,189 @@
+  command explicitly.
+
+### How is the data stored?
+
+The data is stored in columnar format.
+
+#### Traditional Row Stores
+
+Consider a typical CSV file, which represents an example of a row store.
+
+```
+name,age,height
+"Ted",27,6.0
+"Emmanuel",45,5.9
+"Cadence",5,3.5
+```
+
+When this data is stored to disk, you can read an entire record in consecutive order. For example, if
+the first `"` is stored at block 1 on disk, then a sequential scan from 1 assigns the first row of
+data: `"ted",27,6.0`. Keep scanning and you get `\n"Emm`... and so on.
+
+You can use the following query that you can execute very quickly:
+
+```
+SELECT *
+FROM people
+WHERE name == 'Ted'
+```
+
+The database can just scan the first row of data. However, the following is more difficult:
+
+```
+SELECT name
+FROM people
+```
+
+Now the database has to read _all_ of the rows and then select the `name` column, which results in a lot
+more overhead.
+
+#### Columnar Stores
+
+Columnar turns the data sideways. For example, you can make a columnar version of the above data
+and still store it in CSV format:
+
+```
+name,"Ted","Emmanuel","Cadence"
+age,27,45,5
+height,6.0,5.9,3.5
+```
+
+Now let's consider how we can query the data when it's stored this way:
+
+```
+SELECT *
+FROM people
+WHERE name == "ted"
+```
+
+In this case, all the data must be read because the
+`(name, age, height)` is not stored together.
+
+Here's another query:
+
+```
+SELECT name
+FROM people
+```
+
+In this case, only the "name" row needs to be read. All the other lines of the
+file can be skipped.
+
+#### Data partitions
+
+You can improve performance even further by taking advantage of partitions. These are entire files of data that share a value for a column. For example, if everyone in the `people` table lived in `DE`, then you can add that to the filename: `/country=DE/people.csv`.
+
+From there, a query engine would have to know how to read that path and understand that
+all of these people share a country. You can query as follows:
+
+```
+SELECT *
+FROM people
+WHERE country == 'US'
+```
+
+In this case, the query engine no longer even has to read the file. It could just look at the path and realize that there is nothing of interest.
+
+Tables are usually partitioned based on dates; e.g., `submission_date` or `DATE(submission_timestamp)`.
+
+#### Data Ordering (clustering)
+
+Another way to improve query performance is to select a subset of data on a field that the data is ordered by. In BigQuery, this is called "clustering". A clustered field is one which is sorted in the underlying data.
+
+For example, if we wanted to get all of ages greater than age 40 in our table above, we might query like this:
+`SELECT age FROM people WHERE age > 45`
+
+Then we would scan all of the `age` field, starting from `27`, then `45`, then `5`.
+
+However, if we sort the data on that field, our table would look like this:
+\```
```

Comment:
I can't remove these with a suggestion, but I had an extra `\` in here (and in the closing backticks).

Reply:
Please revise as follows:
Then all `age` field are scanned, starting from `27`, proceeding to `45`, and ending with `5`.

If data is sorted based on that field, the table looks like this: